primitives/runtime/src/traits.rs

```diff
@@ -1143,8 +1149,21 @@ pub trait AccountIdConversion<AccountId>: Sized {
 /// Format is TYPE_ID ++ encode(parachain ID) ++ 00.... where 00... is indefinite trailing zeroes to
 /// fill AccountId.
 impl<T: Encode + Decode + Default, Id: Encode + Decode + TypeId> AccountIdConversion<T> for Id {
-	fn into_sub_account<S: Encode>(&self, sub: S) -> T {
-		(Id::TYPE_ID, self, sub).using_encoded(|b|
+	fn into_sub_account<S: Encode>(&self, sub: S) -> Option<T> {
+		if size_of::<T>() < size_of_val(&Id::TYPE_ID) + size_of::<Self>() + sub.size_hint() {
+            None
+        } else {
+		    (Id::TYPE_ID, self, sub).using_encoded(|b|
+			    T::decode(&mut TrailingZeroInput(b))
+		    ).ok()
+        }
+	}
+
+	fn hash_into_sub_account<H: Hash, S: Encode + AsRef<[u8]>>(&self, sub: S) -> T
+		where
+			H::Output: Encode
+	{
+		(Id::TYPE_ID, self, <H as Hash>::hash(sub.as_ref())).using_encoded(|b|
 			T::decode(&mut TrailingZeroInput(b))
 		).unwrap_or_default()
```

Comment:
I know it is not from this PR but the unwrap_or_default here is also suspicious to me.
if `T` doesn't decode any from kind of byte slice then this will fail and always return default without anybody noticing.

I feel we can bring a better trait in order to generate an account id safely.

Reply:
Yes, I also noticed that as a sloppy implementation.